### PR TITLE
fix: retryable client side errors mechanism

### PR DESCRIPTION
**Pull request by kasrakhosravi**
Since when we have client error it'll not be retried properly towards other upstreams, it would be useful to have a mechanism for such retryable client-side errors.

### CODE REVIEW

File: common/errors.go [hunk 1]
```diff
@@ -1958,6 +1968,23 @@ func HasErrorCode(err error, codes ...ErrorCode) bool {
...
+			// If “retriable: true” is present in the details, treat it as "retryable"
+			if details := e.Details(); details != nil {
+				if r, has := details["retriable"].(bool); has && r {
+					return true
+				}
+			}
+			// Otherwise, it’s not retriable
+			return false
+		}
+	}
```
Comment (aramalipoor) [inline common/errors.go:1986]: Probably can be replaced by something like:
```go
if rt, ok := err.DeepSearch("retriable").(bool); ok && !rt {
  return false
}
```

Although the `DeepSearch` needs to be updated to return `interface{}` or `any` meaning any type and not only strings!

Comment (aramalipoor) [reply common/errors.go:1986]: I also suggest to put this lookup 'after' the HasErrorCode logic below, because that one is faster and more often hit vs this one (I'm guessing)

File: architecture/evm/error_normalizer.go [hunk 2]
```diff
@@ -299,6 +299,20 @@ func ExtractJsonRpcError(r *http.Response, nr *common.NormalizedResponse, jr *co
 					nil,
 				)
 			}
+			// Special case for envio's handling of "eth_getBlockReceipts" where they don't support object of
+			// type "BlockNumber" in the "blockHash" field.
+			if strings.Contains(msg, "invalid type: map, expected BlockNumber, 'latest', or 'earliest'") {
```
Comment (aramalipoor) [inline architecture/evm/error_normalizer.go:304]:
```suggestion
			if strings.Contains(msg, "invalid type:") {
```
How about this to make it more generic invalid arg err?

**Comment by aramalipoor** [comment]
@kasrakhosravi the implementation is good imo (with suggested changes) we need to think about all client-side errors that we currently have to decide which ones must be retriable.

The current behavior it retries all client-side errors (check failsafe.go), for example Tenderly returns a 4xx on certain features they don't support which must be retried on other upstreams.

Probably it's good to go over all current instances of ClientSideException and decide one by one if they must be marked as retriable, and then to be safe the default ClientSideException (where we don't know exactly what is the reason) to set it to 'true'. The downside would be extra retries which is better than prematurely failing-fast without trying other upstreams if we really must.

File: architecture/evm/error_normalizer.go (hunk 2 above)
Comment (kasrakhosravi) [reply architecture/evm/error_normalizer.go:304]: I also thought keeping them more specific is better. also not sure if all `invalid type` errors fall under the `retryable` category. needs to be investigated.

File: architecture/evm/error_normalizer.go [hunk 3]
```diff
@@ -276,29 +324,29 @@ func ExtractJsonRpcError(r *http.Response, nr *common.NormalizedResponse, jr *co
...
-			}
+			strings.Contains(msg, "invalid params") ||
+			strings.Contains(msg, "tx of type") ||
+			strings.Contains(msg, "invalid type") {
+
+			// by default, we retry this type of client-side exception, as the root cause
+			// might be an invalid argument, that another upstream might support.
+			//
+			// Examples:
+			// - Envio not supporting { blockNumber: XXX, blockHash: YYY } for eth_getBlockReceipts
```
Comment (aramalipoor) [inline architecture/evm/error_normalizer.go:349]: For majority of the cases this is probably not correct. Validation errors or invalid params can happen a lot if user provides a bad HEX number for example. I suggest to treat these vendor-specific ones explicitly vs retrying all 4xx errors.

A counter-argument is we don't expect properly designed prod dApps to generate these kinds of errors as they'll have higher level validation checks before calling RPC. Although it's in theory, best is to check big RPC clients how many 4xx do they see.

What do you think?

File: architecture/evm/error_normalizer.go [hunk 4]
```diff
@@ -307,14 +355,19 @@ func ExtractJsonRpcError(r *http.Response, nr *common.NormalizedResponse, jr *co
...
+		if code == -32600 {
 			if dt, ok := err.Data.(map[string]interface{}); ok {
-				if msg, ok := dt["message"]; ok {
-					if strings.Contains(msg.(string), "validation errors in batch") {
-						// Intentionally return a server-side error for failed requests in a batch
-						// so they are retried in a different batch.
-						// TODO Should we split a batch instead on json-rpc client level?
+				if innerMsg, ok := dt["message"]; ok {
+					if strings.Contains(innerMsg.(string), "validation errors in batch") {
+						// Return a server-side error so the caller might retry or split the batch.
```
Comment (aramalipoor) [inline architecture/evm/error_normalizer.go:370]: Since we have retryable client errors I think we can remove this edge-case? There's a unit test for that I expect that should pass

File: thirdparty/quicknode.go [hunk 5]
```diff
@@ -60,17 +60,31 @@ func (v *QuicknodeVendor) GetVendorSpecificErrorIfAny(resp *http.Response, jrr i
...
 			)
 		} else if strings.Contains(msg, "failed to parse") {
+			// We do not retry on parse errors, as retrying another upstream would not help.
 			return common.NewErrEndpointClientSideException(
 				common.NewErrJsonRpcExceptionInternal(code, common.JsonRpcErrorParseException, msg, nil, details),
+				false, // not retryable
 			)
-		} else if code == -32010 {
+		} else if code == -32010 { // Transaction cost exceeds current gas limit
+			// We do not retry low gas limit errors, as retrying another upstream would not help.
```
Comment (aramalipoor) [inline thirdparty/quicknode.go:69]: If i'm not wrong the gas limit is a node-specific config, if that's true then trying on another upstream might be actually a correct behavior

Comment (kasrakhosravi) [reply thirdparty/quicknode.go:69]: This indeed seems to be from [Parity](https://github.com/openethereum/parity-ethereum/blob/55c90d4016505317034e3e98f699af07f5404b63/rpc/src/v1/helpers/errors.rs#L452)

Comment (aramalipoor) [reply thirdparty/quicknode.go:69]: I mean the "max gas" is probably a node-specific config (geth, reth, etc), based on this code it seems to be: https://github.com/openethereum/parity-ethereum/blob/v2.7.2-stable/miner/src/pool/verifier.rs#L174

Comment (kasrakhosravi) [reply thirdparty/quicknode.go:69]: seems like it. if I understood correctly, each client can configure their own `gas limit` and also always make sure it is always lower than the `block gas limit overall`.

So one client (e.g reth) might have configured a gas limit higher than parity (but still below block gas limit) and therefore retrying another upstream would make sense under this logic.

File: architecture/evm/error_normalizer.go (hunk 3 above)
Comment (kasrakhosravi) [reply architecture/evm/error_normalizer.go]: totally agreed. I think the invalid arg, params, request is too broad to always retry. would be nice though to get as many of these invalid `arg -> still retryable` error messages as possible.

File: thirdparty/quicknode.go [hunk 6]
```diff
@@ -60,17 +60,32 @@ func (v *QuicknodeVendor) GetVendorSpecificErrorIfAny(resp *http.Response, jrr i
...
+				return common.NewErrEndpointClientSideException(
+					common.NewErrJsonRpcExceptionInternal(code, common.JsonRpcErrorInvalidArgument, msg, nil, details),
+					false, // not retryable
+				)
+			} else if strings.Contains(msg, "eth_getLogs and eth_newFilter are limited") {
+				// retrying on limited range errors for eth_getLogs and eth_newFilter toward other upstreams would be helpful.
+				return common.NewErrEndpointClientSideException(
+					common.NewErrJsonRpcExceptionInternal(code, common.JsonRpcErrorInvalidArgument, msg, nil, details),
+					true, // retryable
+				)
```
[the remark this answers is not in the document]
Comment (kasrakhosravi) [reply thirdparty/quicknode.go]: indeed. there was a mistake in quicknode's error docs: https://www.quicknode.com/docs/base/error-references

sending a large range `eth_getLogs` request to quicknode actually returns another error code:

```
{
    "jsonrpc": "2.0",
    "id": 1,
    "error": {
        "code": -32614,
        "message": "eth_getLogs is limited to a 10,000 range"
    }
}
```

[the remark this answers is not in the document]
Comment (aramalipoor) [reply thirdparty/quicknode.go]: lol, alright let's include both messages then, and both to be TooLarge error

File: architecture/evm/error_normalizer.go (hunk 4 above)
Comment (kasrakhosravi) [reply architecture/evm/error_normalizer.go]: I think the edge-case itself. cannot be removed cause it seems a specific one that carries a certain error in `data.message`. but I created a branch for it in our now organized `Invalid Request` category which would return a client-side exception which is retryable.

**Comment by kasrakhosravi** [comment]
> We have a "ignoreClientErrors" flag in `failsafe.retry` policy, with this change we need to revisit if that flag is needed at all 🤔

don't think we need that flag anymore. and now our code is also cleaner and easier to reason about as we either explicitly define which client errors are retryable and which are not.

File: bin/erpc-server
Comment (aramalipoor) [inline bin/erpc-server:1]: probably we should add these to .gitignore

File: common/errors.go [hunk 7]
```diff
@@ -938,7 +935,10 @@ func (e *ErrUpstreamsExhausted) DeepestMessage() string {
 }
 
 func (e *ErrUpstreamsExhausted) UpstreamId() string {
-	return e.DeepSearch("upstreamId")
+	if val := e.DeepSearch("upstreamId"); val != nil {
+		return val.(string)
```
Comment (aramalipoor) [inline common/errors.go:939]: To avoid unexpected panics maybe:
```suggestion
		if s, ok := val.(string); ok { return s }
```
?

File: architecture/evm/error_normalizer.go [hunk 8]
```diff
@@ -56,7 +56,31 @@ func ExtractJsonRpcError(r *http.Response, nr *common.NormalizedResponse, jr *co
...
+			return common.NewErrEndpointUnsupported(
+				common.NewErrJsonRpcExceptionInternal(
+					int(code),
+					common.JsonRpcErrorUnsupportedException,
+					err.Message,
+					nil,
+					details,
+				),
+			)
+		}
```
Comment (aramalipoor) [inline architecture/evm/error_normalizer.go:78]: In recent commit had to move this to avoid premature detection when message is only "not found" e.g. from Tenderly, we had to move this in different location.

https://github.com/erpc/erpc/pull/204/commits/6df5524559940926102bf3f62ffed6c2e6dae2ce

Comment (kasrakhosravi) [reply architecture/evm/error_normalizer.go]: gotcha. I moved the `unsupported` error normalization logic below `not found` block and also put a comment for why we want to have it there.

File: common/errors.go [hunk 9]
```diff
@@ -1524,12 +1526,22 @@ type ErrEndpointClientSideException struct{ BaseError }
 
 const ErrCodeEndpointClientSideException = "ErrEndpointClientSideException"
 
-var NewErrEndpointClientSideException = func(cause error) error {
+func NewErrEndpointClientSideException(cause error, retryableTowardNetwork ...bool) error {
```
Comment (aramalipoor) [inline common/errors.go:1529]: How about making this an array of flags/options?

```suggestion
func NewErrEndpointClientSideException(cause error, opts ...ExceptionOption) error {
```

then usage will be: 
```golang
common.NewErrEndpointClientSideException(cause, RetryableTowardNetwork)
```

Comment (kasrakhosravi) [reply common/errors.go:1529]: ah gotcha. so more explicit and also if we want to have other flags such as `RetryableTowardUpstream`, we can have that as well.

Comment (kasrakhosravi) [reply common/errors.go:1529]: Exactly. with the logic that we have in `isRetryableTowardNetwork`, if we add it explicitly to any of the exceptions details param; we would explicitly set the retyable behavior.

File: common/errors.go [hunk 10]
```diff
@@ -1945,8 +1967,39 @@ func HasErrorCode(err error, codes ...ErrorCode) bool {
...
+			errs := exher.Errors()
+			if len(errs) > 0 {
+				for _, e := range errs {
+					if IsRetryableTowardsUpstream(e) {
+						return true
+					}
+				}
+			}
+		}
+	}
```
Comment (aramalipoor) [inline common/errors.go:1995]: Shouldn't this be inside IsRetryableTowardsUpstream ?

File: common/errors.go [hunk 11]
```diff
@@ -1973,7 +2024,17 @@ func IsRetryableTowardsUpstream(err error) bool {
...
 		ErrCodeEndpointRequestTooLarge,
-	)
+	) {
+		return false
+	}
+
+	// If the upstream is hitting capacity limits -> no retry
+	if IsCapacityIssue(err) {
+		return false
+	}
```
Comment (aramalipoor) [inline common/errors.go:2034]: Is this a breaking change compared to current behavior?

File: common/errors.go (hunk 10 above)
Comment (kasrakhosravi) [reply common/errors.go:1995]: in the current logic, we are checking this against network scope

https://github.com/erpc/erpc/blob/671178233890f88c0d1918d2b209eeec2bba214b/upstream/failsafe.go#L323

File: common/errors.go [hunk 12]
```diff
@@ -90,13 +90,23 @@ type StandardError interface {
...
 
+type ExceptionOption func(e StandardError)
+
+func WithRetryableTowardNetwork(retryable bool) ExceptionOption {
+	return func(e StandardError) {
+		if e.Base() != nil {
+			e.Base().Details["retryableTowardNetwork"] = retryable
+		}
+	}
+}
```
Comment (aramalipoor) [inline common/errors.go:108]:
```suggestion
func (e BaseError) WithRetryableTowardNetwork(retryable bool) ExceptionOption {
// ..
}
```

If we make this a func of the BaseError then we can easily make any error instance retryable or not, either when initializing such errors or when used...

The usage will be simpler:
```golang
common.NewErrEndpointClientSideException(...).WithRetryableTowardNetwork(true)
```

File: common/errors.go [hunk 13]
```diff
@@ -90,13 +90,23 @@ type StandardError interface {
...
 	Base() *BaseError
 	MarshalZerologObject(v *zerolog.Event)
 }
 
+type ExceptionOption func(e StandardError)
+
+func WithRetryableTowardNetwork(retryable bool) ExceptionOption {
+	return func(e StandardError) {
+		if e.Base() != nil {
+			e.Base().Details["retryableTowardNetwork"] = retryable
```
Comment (aramalipoor) [inline common/errors.go:105]: On a different note, should we make sure Details exist before setting a value?

File: common/errors.go (hunk 11 above)
Comment (kasrakhosravi) [reply common/errors.go:2034]: we have this referenced in two places in the current code and encapsulating it inside the `isRetryableTowardUpstream` would make our code cleaner

here since we are returning `not retryable` for `IsCapacityIssue `, should cover the case here

https://github.com/erpc/erpc/blob/671178233890f88c0d1918d2b209eeec2bba214b/upstream/failsafe.go#L312

and same logic here

https://github.com/erpc/erpc/blob/671178233890f88c0d1918d2b209eeec2bba214b/erpc/networks.go#L279

File: common/errors.go [hunk 14]
```diff
@@ -1945,8 +1967,39 @@ func HasErrorCode(err error, codes ...ErrorCode) bool {
...
+					if IsRetryableTowardsUpstream(e) {
+						return true
+					}
+				}
+			}
+		}
+	}
+
+	// If the error is retryable towards upstream, then it is also retryable towards network
+	return IsRetryableTowardsUpstream(err)
```
Comment (aramalipoor) [inline common/errors.go:1998]: If this is true, then does it mean we're doing a redundant `IsRetryableTowardsUpstream` in failsafe.go, for the flow of network scope? Is there any chance to simplify that?

File: common/errors.go (hunk 11 above)
Comment (aramalipoor) [reply common/errors.go:2034]: Makes sense, I'm wondering why we didn't have it inside before. My guess is because we didn't have network-level retryable logic, am I correct?

File: common/errors.go (hunk 10 above)
Comment (aramalipoor) [reply common/errors.go:1995]: That was probably a different reason (maybe cuz we didn't expect this code for upstream errors).

Logically it does seem like a IsRetryableTowardsUpstream case, what do you think?

File: common/errors.go [hunk 15]
```diff
@@ -1524,12 +1541,15 @@ type ErrEndpointClientSideException struct{ BaseError }
...
 const ErrCodeEndpointClientSideException = "ErrEndpointClientSideException"
 
-var NewErrEndpointClientSideException = func(cause error) error {
+func NewErrEndpointClientSideException(cause error) RetryableError {
 	return &ErrEndpointClientSideException{
-		BaseError{
+		BaseError: BaseError{
 			Code:    ErrCodeEndpointClientSideException,
-			Message: "client-side error when sending request to remote endpoint",
+			Message: "client-side error, etc.",
```
Comment (aramalipoor) [inline common/errors.go:1548]: This message indicates the error is initiated by the upstream, after request was sent to endpoint. The new message is too vague and wouldn't be useful for narrowing down.

In general not sure how to feel about error msgs, maybe we wouldn't really need them... could be a refactor for another day

Comment (kasrakhosravi) [reply common/errors.go:1548]: hmm makes sense. do you think error code is descriptive enough?

File: common/errors.go (hunk 11 above)
Comment (aramalipoor) [reply common/errors.go:2032]: Ah nice find!

File: upstream/failsafe.go [hunk 16]
```diff
@@ -293,47 +293,16 @@ func createRetryPolicy(scope common.Scope, cfg *common.RetryPolicyConfig) (fails
...
-						}
-						if !cfg.IgnoreClientErrors && common.IsClientError(err) {
-							return true
-						}
-					}
-				}
+			return common.IsRetryableTowardsUpstream(err)
+		} else if scope == common.ScopeNetwork && err != nil {
+			if rt := common.IsRetryableTowardNetwork(err); rt != nil {
+				return *rt
```
Comment (aramalipoor) [inline upstream/failsafe.go:305]: I'm not sure if this approach gives us what we need, because in case result is `true` we don't wanna short-circuit and return, we actually want to continue in case for some other reason we should return false (e.g. MissingData is retryable, unless it's from an already synced upstream)

So probably instead of pointer-bool approach it should look like:
```suggestion
			if !common.IsRetryableTowardNetwork(err) {
				return false
```

Although, please go over the full logic to make sure what i say makes sense, because we might want the other way `if common.IsRetryableTowardNetwork(err) { return true }`

File: common/errors.go (hunk 15 above)
Comment (aramalipoor) [reply common/errors.go]: In general I think so, I don't remember any two errors that had same code but diff message, which means even a reference code of Code -> Explanation would be sufficient vs bloating the memory + code with long strings (Imagine on a 10k RPS instance thousands of copies of such long strings are created in memory). The overall mb might be negligible.

Anyway we can think about removing Message concept in a separate refactor if it made sense

File: thirdparty/quicknode.go [hunk 17]
```diff
@@ -60,17 +61,21 @@ func (v *QuicknodeVendor) GetVendorSpecificErrorIfAny(resp *http.Response, jrr i
...
+			).WithRetryableTowardNetwork(false)
+		} else if code == -32010 { // Transaction cost exceeds current gas limit
+			// retrying on gas limit exceeded errors toward other upstreams would be helpful, as max gas limit
+			// can be defined per client (reth, geth, parity, etc.) (still needs to be lower than overall block gas limit)
 			return common.NewErrEndpointClientSideException(
 				common.NewErrJsonRpcExceptionInternal(code, common.JsonRpcErrorClientSideException, msg, nil, details),
 			)
-		} else if code == -32602 {
+		} else if code == -32602 && strings.Contains(msg, "cannot unmarshal hex string") {
+			// we do not retry on invalid argument errors, as retrying another upstream would not help.
```
Comment (aramalipoor) [inline thirdparty/quicknode.go:75]: We probably should separate these two, an 'invalid argument' could be because Quicknode does not support certain param/tag... which we should retry on other nodes.

The 'cannot unmarshal hex string' is most probably safe to not retry.

Comment (kasrakhosravi) [reply thirdparty/quicknode.go:75]: Am I correct that the `invalid argument (32602)` is a bit too general to be considered retryable?

https://github.com/erpc/erpc/pull/203/files#diff-0b005aa95e710cc341ee8cddc3fb690c956286ea19b90299a23e1c416470fceeR366

Comment (kasrakhosravi) [reply thirdparty/quicknode.go:75]: And also the logic here `else if code == -32602 && strings.Contains(msg, "cannot unmarshal hex string")`

is specific enough to be checked here against quicknode (usage of `&`) and if it is just a `32602 `, it will be checked anyway in `error_normalizer`

File: architecture/evm/error_normalizer.go [hunk 18]
```diff
@@ -257,66 +301,78 @@ func ExtractJsonRpcError(r *http.Response, nr *common.NormalizedResponse, jr *co
...
 					}
 				}
 			}
+		} else if code == -32602 || code == -32600 ||
+			strings.Contains(msg, "param is required") ||
+			strings.Contains(msg, "Invalid Request") ||
+			strings.Contains(msg, "validation errors") ||
+			strings.Contains(msg, "invalid argument") ||
+			strings.Contains(msg, "invalid params") ||
+			strings.Contains(msg, "tx of type") {
```
Comment (aramalipoor) [inline architecture/evm/error_normalizer.go:372]: This specific error is probably for tx type support, if yes then we need to retry it on network-level

Comment (aramalipoor) [reply architecture/evm/error_normalizer.go:372]: I see it's already covered above so actually let's remove it from here

File: architecture/evm/error_normalizer.go [hunk 19]
```diff
@@ -257,66 +301,78 @@ func ExtractJsonRpcError(r *http.Response, nr *common.NormalizedResponse, jr *co
...
+		// for certain args/params for certain methods, among different providers or clients. We will retry them.
+		//
+		// Examples:
+		if strings.Contains(msg, "tx of type") || // Should be retried toward upstreams that support this tx type
+			strings.Contains(msg, "invalid type: map, expected BlockNumber, 'latest', or 'earliest'") { // Envio not supporting { blockNumber: XXX, blockHash: YYY } for eth_getBlockReceipts
+			return common.NewErrEndpointClientSideException(
 				common.NewErrJsonRpcExceptionInternal(
 					int(code),
-					common.JsonRpcErrorUnsupportedException,
+					common.JsonRpcErrorInvalidArgument,
```
Comment (aramalipoor) [inline architecture/evm/error_normalizer.go:343]: Still common.JsonRpcErrorUnsupportedException is more accurate is there a reason to make it InvalidArgument?

File: architecture/evm/error_normalizer.go (hunk 18 above)
Comment (kasrakhosravi) [reply architecture/evm/error_normalizer.go]: good catch. fixed

File: architecture/evm/error_normalizer.go (hunk 19 above)
Comment (kasrakhosravi) [reply architecture/evm/error_normalizer.go:343]: hmm I think for the first one (`tx of type`) the `unsupported` makes more sense while for the second one the `invalid argument`. maybe best to separate those here?

File: thirdparty/quicknode.go (hunk 17 above)
Comment (aramalipoor) [reply thirdparty/quicknode.go:75]: I think the code is too generic to not retry (i.e. it's better we retry such generic code) but specific texts about param required/unmarshal errors probably is safe to not retry.

In general when in doubt let's retry.

File: architecture/evm/error_normalizer.go (hunk 19 above)
Comment (aramalipoor) [reply architecture/evm/error_normalizer.go:343]: Even for second is basically a tag that they don't support, so it should be fine imo
